Do you handle multi os dev teams (#9566)
SSWConsulting/SSW.Rules.Content: rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md

---
seoDescription: Learn how to manage a multi-OS development team in source control. This article covers best practices for handling line endings (CRLF vs LF), creating cross-platform setup scripts, configuring Git for consistency, and using .NET CLI for secrets management to boost efficiency across Windows, macOS, and Linux.
type: rule
title: Do you handle Multi-OS dev teams in source control?

> uri: do-you-handle-multi-os-dev-teams-in-source-control

Reviewer: Hi @tomek-i 

As per other rules,

- [ ] 1. Revome the "do-you-" from URLs (includes folder name)

Author: ✅ done

authors:
  - title: Tom Iwainski
    url: https://www.ssw.com.au/people/thomas-iwainski
related: []
created: 2024-11-05T12:52:08.000Z
archivedreason: null

> guid: 7bc2e9c8-2b12-466b-8ecd-34576e3a1648

Reviewer 2: Rule needs to be linked to a category.

Author: Which one?

Author: ✅ done. rules-to-better-code

---

> In today’s development world, teams often consist of developers working on different operating systems (OS), such as Windows, macOS, and Linux. While each OS has its own strengths, managing a cross-platform development environment can introduce challenges. Issues like inconsistent line endings, platform-specific setup scripts, and configuration mismatches can lead to headaches for both individual developers and the team as a whole.

Reviewer: ```suggestion
In today’s development world, teams often consist of developers working on different operating systems (OS), such as Windows, macOS, and Linux. While each OS has its own strengths, managing a cross-platform development environment can introduce challenges. 

Issues like inconsistent line endings, platform-specific setup scripts, and configuration mismatches can lead to headaches for both individual developers and the team as a whole.
```

<!--endintro-->

## Addressing Line Ending differences (CRLF vs. LF)

One of the most common issues faced by teams working with Git across different platforms is the handling of line endings. Windows uses **CRLF** (Carriage Return + Line Feed) for line endings, while macOS and Linux use **LF** (Line Feed) only. This can lead to unnecessary diffs in Git and potential merge conflicts.

### Solution: Use Git’s Line Ending configuration

Git provides a way to manage line endings across different operating systems by using the `core.autocrlf` setting. This configuration ensures that line endings are normalized when files are checked in and out of the repository.

> - **Windows Users**: Set Git to automatically convert line endings to **CRLF** when checking out files, and convert them back to **LF** when committing.

Reviewer: ```suggestion
- **Windows users**: Set Git to automatically convert line endings to **CRLF** when checking out files, and convert them back to **LF** when committing.
```

  ```bash
  git config --global core.autocrlf true
  ```

> - **macOS/Linux Users**: Set Git to automatically convert **CRLF** line endings to **LF** when checking out files and leave them as **LF** when committing.

Reviewer: ```suggestion
- **macOS/Linux users**: Set Git to automatically convert **CRLF** line endings to **LF** when checking out files and leave them as **LF** when committing.
```

  ```bash
  git config --global core.autocrlf input
  ```

> - **Repository-Wide Configuration**: It’s a good practice to enforce this configuration across the team via a `.gitattributes` file, which allows you to define how specific file types should be handled. For example:

Reviewer: ```suggestion
- **Repository-Wide configuration**: It's a good practice to enforce this configuration across the team via a `.gitattributes` file, which allows you to define how specific file types should be handled. For example:
```

  ```text
  * text=auto

>   *.cs text eol=lf

Reviewer 2: I don't think this is right for `*.cs` files. We shouldn't have any specific configuration. I agree with the settings for `*.sh` and `*.bat`.

Author: even for *.cs files I am getting Visual Studio always popping up with the file ending has changed `do you want overwrite it?` and that is usually when our MacOS people touched a file :-D

I can take it out though, let me know

Reviewer 2: I still don't think it's needed. I believe the problem is elsewhere.

Have a read of this post from Scott H: https://www.hanselman.com/blog/youre-just-another-carriage-return-line-feed-in-the-wall

Author: ✅ Done - removed it.

  *.sh text eol=lf
  *.bat text eol=crlf
  ```

This ensures that, no matter what OS a developer is using, files are checked out and committed with consistent line endings. The `eol` attribute specifically handles cases like batch scripts or shell scripts that may need different line endings.

## Creating Cross-Platform Setup Scripts for Easy Onboarding

Onboarding new developers is a critical step in ensuring that everyone is up and running quickly, but multi-OS teams often struggle with platform-specific setup instructions. The goal is to make onboarding as seamless as possible, whether a developer is using Windows, macOS, or Linux.

> ### Solution: Write Cross-Platform setup scripts

Reviewer 2: I think this section needs to be re-written. Powershell is cross platform, and should work across all OS's. Therefore the suggestion should just be 'use powershell' 😄

Author: oOh i thought pwsh was a 3rd party thing that executes powershell scripts :-D

Author: ✅ done

              RunCommand("powershell.exe", "-ExecutionPolicy Bypass -File setup.ps1");
          }
          else if (os.Contains("Unix"))
          {
              RunCommand("bash", "setup.sh");
          }
          else
          {
              Console.WriteLine("Unsupported OS");
          }
      }

      static void RunCommand(string command, string arguments)
      {
          var processStartInfo = new ProcessStartInfo
          {
              FileName = command,
              Arguments = arguments,
              RedirectStandardOutput = true,
              UseShellExecute = false
          };

          using (var process = Process.Start(processStartInfo))
          using (var reader = process.StandardOutput)
          {
              string result = reader.ReadToEnd();
              Console.WriteLine(result);
          }
      }
  }
```

This script checks the platform and runs the appropriate setup script based on the OS.

## Consistent Git configuration for Multi-OS Teams

To further ensure smooth collaboration among multi-OS teams, it’s important to standardize Git configurations. Beyond line endings, Git offers other configurations that help maintain consistency.

> ### Key Git cnfigurations for Multi-OS Teams:

Reviewer 2: ```suggestion
### Key Git configurations for Multi-OS Teams:
```

- **User Name and Email**: Ensure each developer has set up their user name and email, as this is crucial for committing with correct author information:

  ```bash
  git config --global user.name "Your Name"
  git config --global user.email "youremail@example.com"
  ```

- **Global `.gitignore`**: A global `.gitignore` file can help ensure that certain system files (e.g., `Thumbs.db` on Windows or `.DS_Store` on macOS) are ignored across all repositories. You can create and set a global `.gitignore` file using the following command:

  ```bash
  git config --global core.excludesfile ~/.gitignore_global
  ```

  And in `~/.gitignore_global`:

  ```text
  .DS_Store
  Thumbs.db
  ```

- **Hooks and Templates**: Some repositories might require hooks or commit templates to enforce conventions like conventional commit messages or certain commit checks. Using a `.githooks` or `.gitmessage` file in the repository can help maintain consistency across platforms.

---

> ## Handling .NET secrets across platforms

Reviewer 2: I don't think secrets has anything to do with git configuration. Consider removing this section.

Author: It has nothing to do with git configuration. It has a lot to do with multi os dev teams as mac users will use VS code for development and they have to manually set the secrets via command line while in VStudio you can use the context menu on the project in the IDE.  (not sure about rider)

As most people will be using VStudio and then write the documentation targeting this IDE they need to figure out the command - it is much better practice in the onboarding documentation to just write the specific command that the devs can copy and paste.

Reviewer 2: Most MacOS users will use Rider for .NET Development. I don't know anyone who solely uses VS Code for .NET.

What you mention above is correct, but the rule is about source control and I don't think this comment belongs here..

Author: ✅ Done.

People who do not have Rider but a MacOS :-D there are some at SSW

> When working with .NET applications, developers often use the Visual Studio UI for managing secrets. The UI doesn’t work well across different platforms. To ensure consistency, the **CLI** is a better choice for managing secrets in a cross-platform environment.

Author: ```suggestion
When working with .NET applications, developers often use the Visual Studio UI for managing local development secrets. The UI doesn’t work well across different platforms or IDE's like VS Code. To ensure consistency, the **CLI** is a better choice for managing secrets in a cross-platform environment.
```

> ### Solution: Use the .NET CLI for secrets management

Author: ```suggestion
### Solution: Use the .NET CLI for local development secrets management
```

Use the `.NET CLI` to store and retrieve secrets, as this approach works consistently on all platforms:

- **Set a secret**:

  ```bash
  dotnet user-secrets set "MySecret" "secret-value"
  ```

- **Get a secret**:

  ```bash
  dotnet user-secrets get "MySecret"
  ```

By relying on the `.NET CLI`, your team ensures that secrets are handled consistently, regardless of the operating system.

For more information visit [Safe storage of app secrets in development in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-8.0&tabs=linux)

---

> ## General best practices for Cross-Platform efficiency

Reviewer 2: IDE's aren't very related to source control. Consider dropping this section.

Author: ✅ done
